### PR TITLE
SINGA-253 Net converter for caffe model

### PR DESCRIPTION
**Pull request by XiangruiCAI**
Convert caffe model into singa model.
It is a very basic implementation. Now it can convert feed forward net and supports pysinga.
Implementation method:
1. read proto file of caffe model using caffe.proto, and serialize it to string.
2. parse the string use singa model.proto and get the layer config.
3. setup each layer and add it to a feed forward net.

Update and example:
1. put conversion funcions into layer.py and converter.py;
2. add 'caffe' option in cifar example which converts alexnet from
caffe to singa.

### CODE REVIEW

File: python/singa/converter.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+from google.protobuf import text_format
+from singa import layer
+from singa import metric
+from singa import loss
+from singa import net as ffnet
+from .proto import model_pb2
+from .proto import caffe_pb2
+
+
+class caffeConverter:
```
Comment (nudles) [inline python/singa/converter.py]: caffeConverter-->CaffeConverter

File: python/singa/converter.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        return parser_object
+
+    def convert_engine(self, layer_conf, caffe_solver):
+        '''
+        Convert caffe engine into singa engine
+        return:
+            a singa engine string
+        '''
+        caffe_engine = ''
+        solver_mode = caffe_solver.solver_mode
```
Comment (nudles) [inline python/singa/converter.py]: may pass the solver_model directly, i.e. replace caffe _solver->solver_mode

File: python/singa/converter.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        if len(caffe_net.layer):
+            layer_confs = caffe_net.layer
+        elif len(caffe_net.layers):
+            layer_confs = caffe_net.layers
+        else:
+            raise Exception('Invalid proto file!')
+
+        net = ffnet.FeedForwardNet()
+        for i in range(len(layer_confs)):
+            if layer_confs[i].type == 'Data' or layer_confs[i].type == 5:
```
Comment (nudles) [inline python/singa/converter.py]: move data, loss and accuracy out of FeedForwardNet?
Net only consists of layers.

add another function for parsing solver to get loss and metric instances?

File: python/singa/converter.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        '''
+        Create singa net based on caffe proto files.
+            net_proto: caffe prototxt that describes net
+            solver_proto: caffe prototxt that describe solver
+            input_sample_shape: shape of input data tensor
+        return:
+            a FeedForwardNet object
+        '''
+        caffe_net = self.read_net_proto()
+        caffe_solver = self.read_solver_proto()
```
Comment (nudles) [inline python/singa/converter.py]: Maybe better make solver optional. 
For the prediction phase, we only need the net proto.
If it is not available, then just set solver_model to cpu.

File: python/singa/converter.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+from singa import metric
+from singa import loss
+from singa import net as ffnet
+from .proto import model_pb2
+from .proto import caffe_pb2
+
+
+class caffeConverter:
+
+    def __init__(self, net_proto, solver_proto, input_sample_shape):
```
Comment (nudles) [inline python/singa/converter.py]: solver_proto=None
input_sample_shape=None
If the net proto has the input shape, then we don't need to pass in it.

**Comment by XiangruiCAI** [comment]
code updated, thanks!

**Pull request merged**